File: apps/jetbrains-plugin/src/main/kotlin/com/windsurf/controller/services/detection/detectors/CodexDetector.kt
package com.windsurf.controller.services.detection.detectors

import com.windsurf.controller.services.detection.AgentDetector
import com.windsurf.controller.services.detection.DetectionContext
import com.windsurf.controller.services.detection.DetectionResult
import com.windsurf.controller.services.detection.checks.PluginRef
import com.windsurf.controller.services.detection.checks.dirExists
import com.windsurf.controller.services.detection.checks.expandHome
import com.windsurf.controller.services.detection.checks.findInstalledPlugin
import com.windsurf.controller.services.detection.checks.whichBinary

class CodexDetector : AgentDetector {
    override val id = "codex"
    override val name = "Codex"
    override val icon = "codex"

    // OpenAI Codex was integrated natively into the JetBrains AI Assistant
    // plugin from v2025.3 onward. The id `com.intellij.ml.llm` is the AI
    // Assistant plugin id. Third-party Codex plugins (marketplace IDs
    // 28264 / 31307 / 29342) have unconfirmed string ids and are
    // intentionally omitted — the official AI Assistant path covers most
    // installs and the binary + config-dir fallbacks cover the rest.
    private val candidatePlugins = listOf(
        PluginRef(id = "com.intellij.ml.llm", minVersion = "2025.3"),
    )
    private val binaryName = "codex"
    private val configDir = "~/.codex"

    override suspend fun detect(ctx: DetectionContext): DetectionResult? {
        val plugin = findInstalledPlugin(candidatePlugins)
        if (plugin != null) {
            return DetectionResult(extensionId = plugin.id, via = DetectionResult.Via.EXTENSION)
        }
        val binPath = whichBinary(binaryName)
        if (binPath != null) {
            ctx.logger.info("[CodexDetector] binary at $binPath")
            return DetectionResult(extensionId = "__binary__:$binaryName", via = DetectionResult.Via.BINARY)
        }
        if (dirExists(expandHome(configDir))) {
            ctx.logger.info("[CodexDetector] config dir $configDir present")
            return DetectionResult(extensionId = "__config__:$binaryName", via = DetectionResult.Via.CONFIG_DIR)
        }
        return null
    }
}

File: apps/vsc-plugin/src/services/agent-detection/detectors/codex.detector.ts
import { findExtension } from '../checks/extension';
import { whichBinary } from '../checks/binary';
import { dirExists, expandHome } from '../checks/config-dir';
import type { AgentDetector, DetectionContext, DetectionResult } from '../types';

/**
 * Detects OpenAI Codex. Three signals checked in order — first hit wins:
 *
 *   1. The official OpenAI VS Code extension (`openai.chatgpt`,
 *      marketplace name "Codex – OpenAI's coding agent").
 *   2. The `codex` binary on PATH.
 *   3. The `~/.codex/` config directory (created after a successful
 *      `codex login`).
 *
 * Every branch returns `isTerminalAgent: true`. The plugin has no
 * strategy to talk to Codex directly; the mobile app, on seeing the
 * flag, routes the user to `codeam-cli` which handles Codex end-to-end.
 */
export class CodexDetector implements AgentDetector {
  readonly id = 'codex';
  readonly name = 'Codex';
  readonly icon = 'codex';

  private static readonly CANDIDATE_EXTENSION_IDS = ['openai.chatgpt'];
  private static readonly BINARY_NAME = 'codex';
  private static readonly CONFIG_DIR = '~/.codex';

  async detect(ctx: DetectionContext): Promise<DetectionResult | null> {
    const extension = findExtension(CodexDetector.CANDIDATE_EXTENSION_IDS, ctx.extensions);
    if (extension) {
      return {
        installed: true,
        extensionId: extension.id,
        isTerminalAgent: true,
        via: 'extension',
      };
    }

    const binPath = await whichBinary(CodexDetector.BINARY_NAME);
    if (binPath) {
      ctx.log.appendLine(`[CodexDetector] binary at ${binPath}`);
      return {
        installed: true,
        extensionId: `__binary__:${CodexDetector.BINARY_NAME}`,
        isTerminalAgent: true,
        via: 'binary',
      };
    }

    if (await dirExists(expandHome(CodexDetector.CONFIG_DIR))) {
      ctx.log.appendLine(`[CodexDetector] config dir ${CodexDetector.CONFIG_DIR} present`);
      return {
        installed: true,
        extensionId: `__config__:${CodexDetector.BINARY_NAME}`,
        isTerminalAgent: true,
        via: 'config-dir',
      };
    }

    return null;
  }
}
